File: algorithms/2_medium/C/04045_count_robot_groups/solution.c
int countGroups(
    const int *position, const int position_len,
    const int *speed, const int speed_len,
    const int distance
) {
    if (position_len != speed_len || position_len <= 0 || position == NULL || speed == NULL) { return 0; }

    const int len = speed_len;
    int cnt = 1;
    int min_speed = speed[len - 1];
    int prev_pos = position[len - 1];

    for (int i = len - 2; i >= 0; i -= 1) {
        if (prev_pos - position[i] > distance && speed[i] <= min_speed) {
            min_speed = speed[i];
            cnt += 1;
        }

        prev_pos = position[i];
    }

    return cnt;
}

File: algorithms/2_medium/C/04045_count_robot_groups/solution.rs
impl Solution {
    pub fn count_groups(position: Vec<i32>, speed: Vec<i32>, distance: i32) -> i32 {
        if position.is_empty() { return 0; }

        let len = speed.len();
        let mut cnt = 1_i32;
        let mut min_speed = speed[len - 1];
        let mut prev_pos = position[len - 1];

        for i in (0..(len - 1)).rev() {
            if prev_pos - position[i] > distance && speed[i] <= min_speed {
                min_speed = speed[i];
                cnt += 1;
            }

            prev_pos = position[i];
        }

        cnt
    }
}
